### PR TITLE
Improve coverage in X509Certificates library

### PR DESCRIPTION
**Pull request by bartonjs**
I did a pass through the X509Certificates code coverage.  One goal was to not have the tests be missing coverage that the legacy internal test system had, but another was to find completely untested chunks of code.

To that end:
* X509ExtensionsCollection string indexer now has tests
* X509Chain now has ApplicationPolicy-verifying tests
* X509Chain now has CertificatePolicy-verifying tests
* Added tests to X509ChainElementCollection's enumerator

The X509Chain tests found a bug in the Unix implementation, and that is addressed in this change.

### CODE REVIEW

**Comment by stephentoub** [comment]
Test code coverage please

File: src/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs [hunk 1]
```diff
@@ -154,5 +155,136 @@ public static void VerifyExpiration_LocalTime(DateTime verificationTime, bool sh
                 }
             }
         }
+
+        [Fact]
+        public static void BuildChain_WithApplicationPolicy_Match()
+        {
+            using (var msCer = new X509Certificate2(TestData.MsCertificate))
+            {
+                X509Chain chain = new X509Chain();
```
Comment (stephentoub) [inline src/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs]: Even though it's a test, we should ideally be disposing of disposable things like this (unless part of the test is not doing so).

File: src/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs [hunk 2]
```diff
@@ -154,5 +155,136 @@ public static void VerifyExpiration_LocalTime(DateTime verificationTime, bool sh
...
+                chain.ChainPolicy.VerificationFlags =
+                    X509VerificationFlags.AllowUnknownCertificateAuthority;
+
+                chain.ChainPolicy.RevocationMode = X509RevocationMode.NoCheck;
+                chain.ChainPolicy.VerificationTime = cert.NotBefore.AddHours(2);
+
+                bool valid = chain.Build(cert);
+                Assert.False(valid, "Chain built validly");
+
+                Assert.True(chain.ChainElements.Count > 0, "chain.ChainElements.Count > 0");
```
Comment (stephentoub) [inline src/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs]: Consider using Assert.InRange instead.

File: src/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs [hunk 3]
```diff
@@ -154,5 +155,136 @@ public static void VerifyExpiration_LocalTime(DateTime verificationTime, bool sh
...
+
+                Assert.NotSame(cert, chain.ChainElements[0].Certificate);
+                Assert.Equal(cert, chain.ChainElements[0].Certificate);
+
+                X509ChainStatus[] chainElementStatus = chain.ChainElements[0].ChainElementStatus;
+                Assert.True(chainElementStatus.Length >= 1, "chainElementStatus.Length >= 1");
```
Comment (stephentoub) [inline src/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs]: Ditto

File: src/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs [hunk 4]
```diff
@@ -154,5 +155,136 @@ public static void VerifyExpiration_LocalTime(DateTime verificationTime, bool sh
...
+                foreach (X509ChainStatus statusItem in chainElementStatus)
+                {
+                    if (statusItem.Status == X509ChainStatusFlags.NotValidForUsage)
+                    {
+                        foundExpectedFlag = true;
+                        break;
+                    }
+                }
+                
+                Assert.True(foundExpectedFlag, "chain.ChainElements[0].ChainElementStatus contains NotValidForUsage");
```
Comment (stephentoub) [inline src/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs]: Could you use Assert.Any here?

Comment (stephentoub) [inline src/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs]: Oops, Assert.Contains

File: src/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs [hunk 5]
```diff
@@ -154,5 +155,136 @@ public static void VerifyExpiration_LocalTime(DateTime verificationTime, bool sh
...
+                bool foundExpectedFlag = false;
+
+                foreach (X509ChainStatus statusItem in chainElementStatus)
+                {
+                    if (statusItem.Status == X509ChainStatusFlags.NotValidForUsage)
+                    {
+                        foundExpectedFlag = true;
+                        break;
+                    }
+                }
```
Comment (stephentoub) [inline src/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs]: Same comments/questions as on the similar tests earlier

**Comment by stephentoub** [comment]
Some minor comments, but LGTM.

File: src/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs (hunk 4 above)
Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs]: I was explaining my "no" when I realized I could create a custom comparer to solve the "no" problem (X509ChainStatus not having equality defined).

File: src/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs (hunk 2 above)
Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs]: InRange!  I knew there was something, but I wasn't finding it.  Thanks :)

File: src/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs (hunk 1 above)
Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs]: I'm going with brain-o.  It's like a typo, but when you forget that a type is disposable now.

File: src/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs (hunk 4 above)
Comment (stephentoub) [inline src/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs]: :smile: Assert.Contains has an overload that takes a Predicate I believe, so you should be able to just use that.

Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs]: Yeah, that one isn't pretty, but it seems to be the most compact.

(Intentional failure)
```
     System.Security.Cryptography.X509Certificates.Tests.ChainTests.BuildChain_
  WithCertificatePolicy_NoMatch [FAIL]
        Assert.Contains() Failure
        Not found: (filter expression)
        In value:  X509ChainStatus[] [System.Security.Cryptography.X509Certific
  ates.X509ChainStatus, System.Security.Cryptography.X509Certificates.X509Chain
  Status]
```

**Comment by bartonjs** [comment]
The force-push made github UI lose the coverage run: http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/code_coverage_windows_prtest/8/

**Comment by bartonjs** [comment]
Coverage deltas:

|Name|Covered|Uncovered|LineCov|BranchCov|
|---|---|---|---|---|
|System.Security.Cryptography.X509Certificates|38|-38|0.90%|0.90%|
|Internal.Cryptography.Pal.ChainPal|10|-10|5.20%|7.90%|
|Internal.Cryptography.Pal.Native.Helpers|1|-1|1.40%|6.20%|
|Microsoft.Win32.SafeHandles.SafeX509ChainHandle|1|-1|7.20%|0.00%|
|System.Security.Cryptography.X509Certificates.X509ChainElementCollection|3|-3|7.00%|0.00%|
|System.Security.Cryptography.X509Certificates.X509ChainElementEnumerator|14|-14|70.00%|100.00%|
|System.Security.Cryptography.X509Certificates.X509ExtensionCollection|9|-9|17.00%|11.10%|

Everything is delta, so "70.00%" is "The percentage covered went up by 70" (in this case, 0% => 70%).

Gained almost 1% on line coverage, to 83.3% for the library.  A lot of what's left uncovered is behind disabled tests (like the intermittent PFX failures).

**Comment by bartonjs** [comment]
Test Innerloop OSX Debug Build and Test please (it reports 1 "test" failed, but there are no fail reports in the output)

**Comment by stephentoub** [comment]
> it reports 1 "test" failed, but there are no fail reports in the output

```
10:05:22 ./run-test.sh: line 131: 70509 Abort trap: 6           ./corerun xunit.console.netcore.exe $testDllName -xml testResults.xml -notrait category=failing -notrait category=OuterLoop -notrait category=$xunitOSCategory -notrait Benchmark=true
10:05:22 error: One or more tests failed while running tests from 'System.Linq.Parallel.Tests'.  Exit code 134.
```

**Comment by bartonjs** [comment]
@stephentoub Aha, I'll add "failed while" to the list of things I search for.  "[FAIL]" and "failed: [1-9]" are my normal ones.

**Comment by stephentoub** [comment]
Yeah, "abort", "error " (with a space at the end), "seg", and "unhandled" are all good to have in the list.

**Pull request merged**